### ISSUE TITLE
Securing access to repo in workflows

### ISSUE THREAD

**Issue by gabibguti** [rubygems/rubygems#6051]
### Describe the problem as clearly as you can

The GitHub workflows are not restricting the permissions. Restricting permissions in workflows can prevent attackers, once inside your workflow, from editing files and publishing malicious versions for example. Settings least-privilege permissions for GITHUB_TOKEN in each workflow can help increase the repository's security against some supply-chain attacks.

This setting is recommended by [GitHub](https://docs.github.com/en/actions/security-guides/automatic-token-authentication) and other security tools, such as [Scorecards](https://docs.github.com/en/actions/security-guides/automatic-token-authentication) and [StepSecurity](https://docs.github.com/en/actions/security-guides/automatic-token-authentication).

----

_This is not a code bug or question report so the other questions are not related._

----

### Additional context

Hi! I am Gabriela and I work on behalf of Google and the OpenSSF suggesting improvements to open-source projects on supply-chain security. rubygems is an important repository for the Ruby community, plus it is considered critical by the OpenSSF.

If you are interested, please let me know, I can open a PR :)

**Comment by deivid-rodriguez** [comment]
Hi @gabibguti! Thanks for reaching out!

We don't trigger workflows on the `pull_request_target` event, so I believe we're not vulnerable to the kind of thing described at https://securitylab.github.com/research/github-actions-preventing-pwn-requests/.

That said, I guess there could be other issues and we may start reacting to that event in the future and forget to consider this.

It doesn't hurt to stay secure by default, so please send us a PR :)

**Comment by gabibguti** [comment]
Hi @deivid-rodriguez!

> We don't trigger workflows on the pull_request_target event, so I believe we're not vulnerable to the kind of thing described at https://securitylab.github.com/research/github-actions-preventing-pwn-requests/.

Ah yes! That's true, you are currently safe against that :) `pull_request_target` can be an entrance for an attacker indeed. Another way would be to do `uses: my-action@v1` and `my-action` is or becomes not trustable.

Anyway, as you said "it doesn't hurt to stay secure by default". I will open a PR then :)

**Comment by gabibguti** [comment]
Hi @deivid-rodriguez!

Do you have default workflow permissions set to "restrictive" in this repo?

If yes, great news! No need to do anything else on this matter.

If not, you can set it up [following this configuration](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository#configuring-the-default-github_token-permissions) and we won't need a PR.

-----

Why is that? Because taking a look at your workflows, most of them perform checkout, setup ruby and test or build. The only permission they need is reading the repository. In other words, setting `contents: read` and everything else to `none`.

This can be done setting permissions in each workflow, which would need a PR, or by changing the default workflow permissions to "restrictive" which is the same thing.

I have tested setting permissions in each workflow in [my fork](https://github.com/gabibguti/rubygems/actions) and seems like everything is working as it should.

If, by any other reason, you still want the PR, I can open it anyways :)

**Comment by simi** [comment]
> This can be done setting permissions in each workflow, which would need a PR, or by changing the default workflow permissions to "restrictive" which is the same thing.

@gabibguti, by `"restrictive" default workflow permission` do you mean `Read repository contents permission` option in `Workflow permissions` section of Actions settings?

**Comment by gabibguti** [comment]
@simi Yes! And the "Allow GitHub Actions to create and approve pull requests" can be unchecked. It should not be needed in this case.

**Comment by hsbt** [comment]
I'll add `ossf/scorecards` action to this repo same as https://github.com/ruby/ruby/pull/6716
